source/simulator/core.cc

```diff
@@ -1678,19 +1678,38 @@ namespace aspect
             free_surface->execute ();
 
           assemble_advection_system (AdvectionField::temperature());
-          build_advection_preconditioner(AdvectionField::temperature(),
-                                         T_preconditioner);
-          solve_advection(AdvectionField::temperature());
+          if (system_rhs.block(AdvectionField::temperature().block_index(introspection)).l2_norm() <= std::numeric_limits<double>::min())
```

Comment:
Okay, I added the comment, and after I noticed that I would have to add it in about 20 places, I moved all my new additions into `solve_advection` so that we do not have to duplicate the code.